source/agora/consensus/EnrollmentManager.d

```diff
@@ -682,6 +638,70 @@ public class EnrollmentManager
         }
     }
 
+    /***************************************************************************
+
+        Generate and store pre-images for this cycle, as well as sparsely
+        selected preimages for other cycles.
+
+        This generates all pre-images needed for a cycle in the range of less
+        than two times of a validator cycle and store them in `cycle_preimages`.
+        Additionally, It generates and stores pre-images being used for defined
+        number of cycles in `preimage_rounds` when there is no pre-image needed.
+        This function is very expensive, but should be seldom called, and might
+        take more than 30ms for generating a hundred thousand number of them.
+
+        Params:
+            height = block height used to determine which range of
+                pre-images will be generated.
+
+    ***************************************************************************/
+
+    private void generatePreimages (ulong height) @safe nothrow
+    {
+        // This determines which range of preimages must be generated.
+        ulong start_height =
+            ((height / ValidatorSet.ValidatorCycle) + 2) *
+                ValidatorSet.ValidatorCycle;
+        ulong bulk_index = start_height / AllCountPreimages;
+        const ulong MaxHeight = (bulk_index + 1) * AllCountPreimages;
+
+        // Clear if recreating pre-images is needed
+        () @trusted {
+            if (this.preimage_rounds.length > 0 &&
```

Comment:
Look at the documentation: You can remove the `.length` call by using a seed

Reply:
> Look at the documentation: You can remove the .length call by using a seed

The `.length` call is for determining whether there are a bulk of pre-images already calculated. I don't understand the relation between removing the `.length` call and using a seed.

Reply:
When I say `seed`, it's a reference to the argument `maxElement` can take. If you seed it with a value, it means it will have  at least this value. However if the range is empty, you know it will have this value.
Hence just doing `if (this.preimage_rounds.byKey.maxElement(0) < start_height)` allows you to remove the `.length > 0` check because the condition will be `false` when `length == 0`.

Reply:
Ah, you meant the documentation for the `maxElement`. I will read and apply what you suggested.